util/singleflight: add DoChanContext

andrew-d opened the pull request:
This is a variant of DoChan that supports context propagation, such that the context provided to the inner function will only be canceled when there are no more waiters for a given key. This can be used to deduplicate expensive and cancelable calls among multiple callers safely.

Updates #11935


Change-Id: Ibe1fb67442a854babbc6924fd8437b02cc9e7bcf

dsnet commented on util/singleflight/singleflight.go:
```diff
@@ -143,6 +150,86 @@ func (g *Group[K, V]) DoChan(key K, fn func() (V, error)) <-chan Result[V] {
+	}
+	// Create a context that is not canceled when the parent context is,
+	// but otherwise propagates all values.
+	callCtx, callCancel := context.WithCancel(context.WithoutCancel(ctx))
```
Whatever semantic we choose, this needs to be documented.

andrew-d replied:
I changed to using `context.Background` and documented it; we might want to change at some point, but IMO it's easier to require explicit value propagation.

dsnet commented on util/singleflight/singleflight.go:
```diff
@@ -143,6 +150,86 @@ func (g *Group[K, V]) DoChan(key K, fn func() (V, error)) <-chan Result[V] {
+// passed to the fn function is a context that is canceled only when there are
+// no callers waiting on a result (i.e. all callers have canceled their
+// contexts).
+func (g *Group[K, V]) DoChanContext(ctx context.Context, key K, fn func(context.Context) (V, error)) <-chan Result[V] {
```
Comment that the returned `Result.Err` is set to `context.Context.Err` if the provided context is canceled.

dsnet commented on util/singleflight/singleflight.go:
```diff
@@ -143,6 +150,86 @@ func (g *Group[K, V]) DoChan(key K, fn func() (V, error)) <-chan Result[V] {
+	// finished executing after the last caller has returned.
+	if c.ctxWaiters.Add(-1) == 0 {
+		c.cancel()
+		c.wg.Wait()
```
If we have 10 concurrent waiters, I find it odd that 9 return immediately, but then the 10th must wait, but maybe that's the right semantic?

dsnet commented on util/singleflight/singleflight.go:
```diff
@@ -143,6 +150,86 @@ func (g *Group[K, V]) DoChan(key K, fn func() (V, error)) <-chan Result[V] {
+	case <-ctx.Done():
+		err = ctx.Err()
+	case res = <-result:
+	}
```
If there's race between the context being canceled and the result being produced, should one take precedence? Alternatively, if the context was already canceled and a result happened to be ready now, what should take precedence?

If we want cancelation to take precedence, we could do:
```go
select {
case <-ctx.Done():
case res = <-result:
}

...

if err := ctx.Err(); err != nil {
	res = Result[V]{Err: err}
}
output <- res
```
Thus, context errors take precedence over results.

andrew-d replied:
I think preferring context cancelation makes sense; if the caller gets a value, it's likely that it'll just end up continuing until a later point and then checking cancelation and aborting, which is essentially wasted work.

dsnet commented on util/singleflight/singleflight_test.go:
```diff
@@ -321,3 +322,153 @@ func TestPanicDoSharedByDoChan(t *testing.T) {
+
+	t.Run("AllCancel", func(t *testing.T) {
+		for _, n := range []int{1, 2, 10} {
+			n := n
```
I don't think this is necessary in Go 1.22, which what our `go.mod` specifies.

https://tip.golang.org/doc/go1.22#language

dsnet commented on util/singleflight/singleflight_test.go:
```diff
@@ -321,3 +322,153 @@ func TestPanicDoSharedByDoChan(t *testing.T) {
+					case <-ctx.Done():
+						return 0, ctx.Err()
+					case <-trigger:
+						t.Fatal("unexpected trigger; want all callers to cancel")
```
[Per `testing.T` docs](https://pkg.go.dev/testing#T):
> A test ends when its Test function ... calls ... Fatal ... Those methods, as ... must be called only from the goroutine running the Test function.

Thus, calling `Fatal` here is invalid since `fn` is run in a goroutine.

dsnet commented on util/singleflight/singleflight_test.go:
```diff
@@ -321,3 +322,153 @@ func TestPanicDoSharedByDoChan(t *testing.T) {
+		ch := g.DoChanContext(ctx, "key", func(_ context.Context) (int, error) {
+			gotVal, ok := ctx.Value(key).(string)
+			if !ok {
+				t.Fatalf("expected value to be present in context")
```
Should be `t.Error`. See my comment below about concurrent `t.Fatal` calls.

dsnet commented on util/singleflight/singleflight_test.go:
```diff
@@ -321,3 +322,153 @@ func TestPanicDoSharedByDoChan(t *testing.T) {
+				t.Fatalf("expected value to be present in context")
+			}
+			if gotVal != value {
+				t.Fatalf("unexpected value; got %q, want %q", gotVal, value)
```
Should be `t.Error`. See my comment below about concurrent `t.Fatal` calls.

dsnet commented on util/singleflight/singleflight_test.go:
```diff
@@ -321,3 +322,153 @@ func TestPanicDoSharedByDoChan(t *testing.T) {
+					chs     []<-chan Result[int]
+					cancels []context.CancelFunc
+				)
+				for i := 0; i < n; i++ {
```
`for i := range n`

dsnet commented on util/singleflight/singleflight_test.go:
```diff
@@ -321,3 +322,153 @@ func TestPanicDoSharedByDoChan(t *testing.T) {
+					chs     []<-chan Result[int]
+					cancels []context.CancelFunc
+				)
+				for i := 0; i < n; i++ {
```
In this test you synchronously call `DoChanContext` N times, and then cancel them N times.

What happens when you interleave `DoChanContext` calls with the cancels?

andrew-d replied:
Good idea; switched to canceling every third goroutine in this loop, and added another test case with 20 waiters to ensure we're exercising enough of this logic.

andrew-d commented:
@dsnet because I'm always a bit scared of heavily-concurrent code... mind taking another look at this before I merge? I think I addressed all your comments and the tests still pass with `-race`, but still 😅
